include/swift/Runtime/MutexPThread.h

```diff
@@ -0,0 +1,102 @@
+//===--- MutexPThread.h - ---------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Mutex, Condition, Read/Write lock, and Scoped lock implementations
+// using PThreads.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_MUTEX_PHTREAD_H
+#define SWIFT_RUNTIME_MUTEX_PHTREAD_H
+
+#include <pthread.h>
+
+namespace swift {
+
+/// Private PThread implementation of Condition functionality.
+/// Use Condition instead (see Mutex.h).
+class ConditionImpl {
+  friend class MutexImpl;
+  friend class Condition;
+
+  ConditionImpl(const ConditionImpl &) = delete;
+  ConditionImpl &operator=(const ConditionImpl &) = delete;
+  ConditionImpl(ConditionImpl &&) = delete;
+  ConditionImpl &operator=(ConditionImpl &&) = delete;
+
+public:
+  ~ConditionImpl();
+
+public:
+  void notifyOne();
+  void notifyAll();
+
+private:
+  explicit ConditionImpl(bool reportFatal);
+  bool ReportFatal;
+  pthread_cond_t PThreadCond;
+};
+
+/// Private PThread implementation of Mutex functionality.
+/// Use Mutex instead (see Mutex.h).
+class MutexImpl {
+  friend class Mutex;
+
+  MutexImpl() = delete;
+  MutexImpl(const MutexImpl &) = delete;
+  MutexImpl &operator=(const MutexImpl &) = delete;
+  MutexImpl(MutexImpl &&) = delete;
+  MutexImpl &operator=(MutexImpl &&) = delete;
+
+public:
+  ~MutexImpl();
+
+public:
+  void lock();
+  void unlock();
+  bool try_lock();
+  void wait(ConditionImpl &condition);
+
+private:
+  explicit MutexImpl(bool reportFatal, bool checked);
+  bool ReportFatal;
```

Comment:
How important is the savings of moving the ReportFatal boolean to a non-type template parameter. In the runtime I see very few instances of mutex, etc. being created. Of course that could change if some aspect of the code for example decides to have a mutex per some swift object, etc... not sure how likely that is.

I ask because adding a template will cause some ripple effects in what I have implemented that I need to get some time to think about how best to manage. 

ReportError is only need for a specific pathway (not likely to change) so stepping back a bit I think I can avoid the template ripple and the boolean storage by providing a less generic way for this one pathway to avoid the fatalReport "looping" issue. I will try to get time a little later today to make this change, should have likely done it this "one of way" initially.